File: CSCenter/AlgorithmicToolbox/Week4/inversions/inversions.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

uint64_t inversions(int *a, int n);
void merge_inv(int *a, int *aux, int lo, int hi, uint64_t *inv_count);
void merge(int *a, int *aux, int lo, int mid, int hi, uint64_t *inv_count);

int main() {

	int n;
	if(scanf("%d", &n) != 1) {
		return 1;
	}
	
	int *a = malloc(sizeof(int) * n);
	for(int i = 0; i < n; i++) {
		if(scanf("%d", &a[i]) != 1) {
			return 1;
		}
	}
	
	printf("%llu\n", inversions(a, n));
	
	free(a);
	return 0;
}

uint64_t inversions(int *a, int n) {
	
	uint64_t inv_count = 0;
	int *aux = malloc(sizeof(int) * n);
	merge_inv(a, aux, 0, n, &inv_count);
	free(aux);
	return inv_count;
}

void merge_inv(int *a, int *aux, int lo, int hi, uint64_t *inv_count) {
	
	if(hi - lo <= 1) {
		return;
	}
	
	int mid = lo + (hi - lo) / 2;
	
	merge_inv(a, aux, lo, mid, inv_count);
	merge_inv(a, aux, mid, hi, inv_count);
	
	merge(a, aux, lo, mid, hi, inv_count);
}

void merge(int *a, int *aux, int lo, int mid, int hi, uint64_t *inv_count) {
	
	for(int i = lo; i < hi; i++) {
		aux[i] = a[i];
	}
	
	int lp = lo, rp = mid, i = lo;
	while(lp < mid && rp < hi) {
		a[i++] = (aux[lp] <= aux[rp]) ? aux[lp++] : ((*inv_count) += (mid - lp), aux[rp++]);
	}
	
	while(i < hi) {
		a[i++] = (lp < mid) ? aux[lp++] : aux[rp++]; 
	}
}

File: Stanford/Inversions/Inversions.java
class Inversions {
	
	private static long inv;
	
	public static long naiveCount(int[] a) {
		inv = 0;
		for(int i = 0; i < a.length; i++) {
			for(int j = i + 1; j < a.length; j++) {
				if(a[i] > a[j]) {
					inv++;
				}
			}
		}
		return inv;
	}
	
	public static long count(int[] a) {
		inv = 0;
		int[] aux = new int[a.length];
		count(a, 0, a.length, aux);
		return inv;
	}
	
	private static void count(int[] a, int lo, int hi, int[] aux) {
		if(hi - lo <= 1) {
			return;
		}
		int mid = lo + (hi - lo) / 2;
		count(a, lo, mid, aux);
		count(a, mid, hi, aux);
		merge(a, lo, mid, hi, aux);
	}
	
	private static void merge(int[] a, int lo, int mid, int hi, int[] aux) {
		for(int i = lo; i < hi; i++) {
			aux[i] = a[i];
		}
		int i = lo, j = mid, k = lo;
		while(i < mid && j < hi) {
			if(aux[i] <= aux[j]) {
				a[k++] = aux[i++];
			} else {
				a[k++] = aux[j++];
				inv += (mid - i);
			}
		}
		while(i < mid) {
			a[k++] = aux[i++];
		}
		while(j < hi) {
			a[k++] = aux[j++];
		}
	}
}